### PR TITLE
Feature/fix gec predictor

### PR DESCRIPTION
**Pull request by Frost45**
In this PR, code changes have been made to the following files:

1. `datareader.py`
2. `gec_predictor.py`
3. `seq2labels_model.py`
4. `test_seq2labels.py`
5. `regression_tests/test_regression_data_predictor.py`

Please review the above files.

I've added additional test cases to:
1. `test_gec_model.py`
2. `test_gec_predictor.py`

I've left in the comments that I added to `gec_model.py` for understanding what was happening.

I have also rebased the `master` branch onto this branch. 

How to test:
- `pytest`
- `python regression_tests/test_regression_data_predictor.py`

### CODE REVIEW

File: gector/gec_predictor.py [hunk 1]
```diff
@@ -143,15 +153,77 @@ def predict_batch_instance(self, instances: List[Instance]) -> List[JsonDict]:
...
+        prev_preds_dict = {
+            id: [final_batch[id].fields["tokens"].tokens]
+            for id in range(len(final_batch))
+        }
+
+        short_ids = [
+            id
+            for id in range(len(final_batch))
+            if len(final_batch[id].fields["tokens"].tokens) < 4
+        ]
```
Comment (desilinguist) [inline gector/gec_predictor.py:171]: It seems a little wasteful to iterate over `final_batch` twice? Couldn't we replace these two statements with a simple for loop and do both things in the same loop?

File: gector/gec_predictor.py [hunk 2]
```diff
@@ -143,15 +153,77 @@ def predict_batch_instance(self, instances: List[Instance]) -> List[JsonDict]:
...
+                elif (
+                    orig.fields["tokens"].tokens != pred.fields["tokens"].tokens
+                    and pred.fields["tokens"].tokens in prev_preds
+                ):
+                    # update final batch, but stop iterations
+                    final_batch[pred_id] = pred
+                else:
+                    continue
+            pred_ids = new_pred_ids
+        return sanitize(final_outputs)
```
Comment (desilinguist) [inline gector/gec_predictor.py:226]: I don't see any comments at all in the code we added. Can we add some please?

**Review by desilinguist — APPROVED** [review]
Thanks for addressing my comments! LGTM!

File: gector/datareader.py [hunk 3]
```diff
@@ -117,17 +134,24 @@ def extract_tags(self, tags: List[str]):
...
+        words: List[str] = None,
+    ) -> Instance:  # type: ignore
         """
         We take `pre-tokenized` input here, because we don't have a tokenizer in this class.
         """
         # pylint: disable=arguments-differ
         fields: Dict[str, Field] = {}
-        sequence = TextField(tokens, self._token_indexers)
+        # Set size of tokens to _max_len + 1 since $START token is being added
+        sequence = TextField(tokens[: self._max_len + 1], self._token_indexers)
```
Comment (ksteimel) [inline gector/datareader.py:154]: Ah! nice find!

File: gector/gec_model.py [hunk 4]
```diff
@@ -149,46 +167,61 @@ def predict(self, batches):
...
 
     def _get_embbeder(self, weigths_name, special_tokens_fix):
-        embedders = {'bert': PretrainedBertEmbedder(
-            pretrained_model=weigths_name,
-            requires_grad=False,
-            top_layer_only=True,
-            special_tokens_fix=special_tokens_fix)
+        embedders = {
+            "bert": PretrainedBertEmbedder(
+                pretrained_model=weigths_name,
```
Comment (ksteimel) [inline gector/gec_model.py:204]: maybe we should correct `weigths_name` to `weights_name`

File: gector/gec_predictor.py [hunk 5]
```diff
@@ -97,6 +105,8 @@ def predict_instance(self, instance: Instance) -> JsonDict:
 
         Parameters
         ---------
+        instance: Instance
```
Comment (ksteimel) [inline gector/gec_predictor.py:108]: thanks for adding this!

File: gector/gec_predictor.py [hunk 6]
```diff
@@ -88,7 +94,9 @@ def predict_batch(self, sentences: List[str]) -> List[JsonDict]:
                 - corrected_words
             For an explanation of each of these see `Seq2Labels.decode()`.
         """
-        return self.predict_batch_json([{"sentence": sentence} for sentence in sentences])
+        return self.predict_batch_json(
+            [{"sentence": sentence} for sentence in sentences]
+        )
 
     @overrides
     def predict_instance(self, instance: Instance) -> JsonDict:
```
Comment (ksteimel) [inline gector/gec_predictor.py:102]: Predict instance should probably create a batch from a single instance and call the revised predict_batch_instance method you've written. Currently, I think batch prediction and single instance prediction (what would most likely be used in the engine) would give you differing results.

File: gector/gec_predictor.py [hunk 7]
```diff
@@ -143,15 +157,102 @@ def predict_batch_instance(self, instances: List[Instance]) -> List[JsonDict]:
...
+            for op_ind, pred_id in enumerate(pred_ids):
+
+                # Update final outputs
+                final_outputs[pred_id] = outputs[op_ind]
+                orig = final_batch[pred_id]
+
+                # Create tokens from corrected words for next iter
+                tokens = [
+                    Token(word)
+                    for word in ["$START"] + outputs[op_ind]["corrected_words"]
```
Comment (ksteimel) [inline gector/gec_predictor.py:219]: If I was betting, I'd say that this is where the discrepancies were coming from.

File: gector/gec_predictor.py [hunk 8]
```diff
@@ -143,15 +157,102 @@ def predict_batch_instance(self, instances: List[Instance]) -> List[JsonDict]:
...
+                    new_pred_ids.append(pred_id)
+                    # Append new prediction to prev_preds_dict
+                    prev_preds_dict[pred_id].append(
+                        pred.fields["tokens"].tokens
+                    )
+                # If model output is same as that in prev iter, update final batch
+                # but stop passing it into the model for future iters
+                # This means that no corrections have been made in this iteration
+                elif (
+                    orig.fields["tokens"].tokens != pred.fields["tokens"].tokens
```
Comment (ksteimel) [inline gector/gec_predictor.py:244]: Oh I see so there was also an "early stopping" component as well where it stopped iterating if no corrections had been made. That's probably also the source of some of the discrepancies.

File: gector/gec_predictor.py [hunk 9]
```diff
@@ -143,15 +157,102 @@ def predict_batch_instance(self, instances: List[Instance]) -> List[JsonDict]:
...
+                    and pred.fields["tokens"].tokens in prev_preds
+                ):
+                    # update final batch, but stop iterations
+                    final_batch[pred_id] = pred
```
Comment (ksteimel) [inline gector/gec_predictor.py:248]: So we set final_batch[i] in cases where i was pointing to an instance with 4 or fewer words, but it doesn't seem like those cases are skipped here. Wouldn't that cause final_batch[i] to be overwritten for these short cases? Is this an error in gec_model that we're replicating for consistency?

**Review by ksteimel — APPROVED** [review]
Thanks for fixing these discrepancies! I'm sure this wasn't easy to track down! 
I do wish stuff like appending $START was done in the dataset reader's text_to_instance method. That'd be the more allennlp-y way to handle it. However, this would break the existing gec_model.py code which is expecting to have to add $START so that doesn't really seem possible right now.

File: gector/gec_model.py (hunk 4 above)
Comment (Frost45) [reply gector/gec_model.py:204]: Yeah, it's a real eyesore.

File: gector/gec_predictor.py (hunk 7 above)
Comment (Frost45) [reply gector/gec_predictor.py:219]: Yeah, it was a combination of things, a lot of it to do with indices and START tokens etc.

File: gector/gec_predictor.py (hunk 8 above)
Comment (Frost45) [reply gector/gec_predictor.py:244]: Exactly! Passing it into the model again seemed to force it to make unnecessary corrections. That's where all the weird sentences I showed you in our meeting came from.

File: gector/gec_predictor.py (hunk 9 above)
Comment (Frost45) [reply gector/gec_predictor.py:248]: This is not the case for short sentences. Such sentences are not passed into the model at all. GEC_Model has a min_length of 3.

This is if no corrections have been made to the sentence in this iteration. So the list of `pred_ids` is not updated for these sentences.

**Comment by Frost45** [comment]
> Thanks for fixing these discrepancies! I'm sure this wasn't easy to track down! I do wish stuff like appending $START was done in the dataset reader's text_to_instance method. That'd be the more allennlp-y way to handle it. However, this would break the existing gec_model.py code which is expecting to have to add $START so that doesn't really seem possible right now.

@ksteimel When we add ensemble prediction to the predictor, we can get rid of GEC_model entirely and make this change.

**Review by damien2012eng — APPROVED** [review]

**Review by desilinguist — APPROVED** [review]

**Pull request merged**